daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/TransactionPreprocessor.scala

```diff
@@ -44,14 +45,26 @@ private[preprocessing] final class TransactionPreprocessor(
   // Translate a GenNode into an expression re-interpretable by the interpreter
   @throws[PreprocessorException]
   def unsafeTranslateNode[Cid <: Value.ContractId](
-      node: Node.GenNode.WithTxValue[Transaction.NodeId, Cid],
-  ): speedy.Command = {
+      acc: (Set[Value.ContractId], Set[Value.AbsoluteContractId]),
+      node: Node.GenNode.WithTxValue[Transaction.NodeId, Cid]
+  ): ((Set[Value.ContractId], Set[Value.AbsoluteContractId]), speedy.Command) = {
+
+    val (localCids, globalCids) = acc
 
     node match {
       case Node.NodeCreate(coid @ _, coinst, optLoc @ _, sigs @ _, stks @ _, key @ _) =>
         val identifier = coinst.template
         val arg = unsafeAsValueWithAbsoluteContractIds(coinst.arg.value)
-        commandPreprocessor.unsafePreprocessCreate(identifier, arg)
+        coid match {
+          case acoid: AbsoluteContractId =>
+            if (globalCids(acoid))
+              fail(s"created contract $coid in is not fresh")
+          case _ =>
+        }
```

Comment:
I'm not really sure why we need this check here. The created contract ID is not really an input to the interpretation, but an output. AFAICS this method is called in `Engine.reinterpret` and `Engine.validate` to translate the nodes into commands. Accordingly, we only need to check freshness of contract IDs that actually make it into the commands and the ID of the created contract will not.

In that sense, the created contract ID should be treated the same way as an exercise result: We don't look at the contract IDs in there either.

Reply:
For Canton, it does not really matter because Canton only uses `reinterpret` and in that case `globalCids` is always empty in this line.

Reply:
The problem is when you have to translate several nodes at the time.
You need to keep track of the contract create in node, that are used as input in subsequent  nodes. Those cids should not be considered as inputs. 

In the current state (where we provide only reinterpretation of root nodes or reinterpretation of a single node), the only case where this could happen is for `CreateAndExercise` command.

Reply:
by the way, I use the renaming `global` -> `input`, `local` -> `output`

Reply:
`input` and `output` are already defined in the ledger model. `output` has a slightly different meaning: https://docs.daml.com/concepts/ledger-model/ledger-integrity.html#internal-consistency
The difference is that if a contract is created and archived in the same transaction, then it's not an output of the transaction.

Reply:
`input` also has a different meaning. You use `input` for any referenced contract whereas `input` in the ledger model means that the contract is actually used (fetch/exercise/...) and not just mentioned in one of the values.

Reply:
I actually like `global` and `local`. Why do you want to rename them?

Reply:
Thanks. Will revert the change.

Reply:
Precisely, I wondered about the case when multiple nodes are translated. There is no guarantee that the contract ID in the create node is the one that will be recomputed. So if someone supplies a wrong contract ID here, then exclude this contract ID from the freshness check in the subsequent commands. However, during interpretation, we will end up with a different discriminator here, which may actually clash with one of the global contract IDs used in the arguments of one of the commands. Unless I missed it, the current freshness checks do not catch this at the moment. However, this can only happen if we're running `validate` (because `reinterpret` takes only at a single node). Fortunately, `validate` checks that the reconstructed transaction replays the given one and will therefore detect that a different contract ID has been computed.

Therefore, IMO this check is pointless here. (That's also why I wrote in my summary that this only works because of the constraints how interpretation is used.)

Reply:
Indeed, the check does not buy you much for the freshness check point of view. 
However, it makes easy to enforce the invariant   `(localCids & globalCids).isEmpty` when folding a list of nodes using `unsafeTranslateNode`. I find having this invariant make reasoning simpler, and the only cost of it is the early rejection bogus transaction.

Reply:
Futhremore, in my point of view, we have to always check the created contract have the expected discriminator.

Reply:
In fact, to be correct we have to ensure the cid created by the transaction generated by the reinterpretation are the same as the one computing during the nodes translation.
As you say, this is enforce by `replay` check.

Reply:
If the invariant helps, then leave the check in. As I wrote, it does not really matter for Canton (and I don't know about Corda). I saw that you replaced the check with `localCids`, which wouldn't enforce this invariant.

A related observation: The check that the recomputed contract IDs are the same doesn't really play well with suffixing of contract IDs. Basically, one has to validate the nodes before suffixing as otherwise the reinterpreted transaction will miss the suffixes. That's not a big deal, but something to keep in mind. (In Canton we reinterpret suffixed nodes and suffix the reinterpreted transaction again before we compare them.)

Reply:
Thanks for catching the typo in the check with `localCids` that should be `globalCids`.

Reply:
You are right the replay should be done while ignoring suffix for local contract ID.
I should probably do that in an upcoming PR.